enhancements/cluster-logging/elasticsearch-cert-management.md

```diff
@@ -0,0 +1,158 @@
+
+We will be successful when:
+
+* Users of Elasticsearch Operator no longer need to maintain certificates for Elasticsearch and can have certificates injected into their secrets
+
+* There is a clear path forward for automatic upgrades
+
+### Non-Goals
+
+* This does not seek to implement work to adhere to the cluster level TLS definitions, however this does make the implementation less complex for the Elasticsearch use-case.
+
+## Proposal
+
+Instead of the current implementation where any consumer of the Elasticsearch Operator is responsible for providing certificates and secrets for Elasticsearch and the operator to use, we want to shift this responsibility to the Elasticsearch Operator instead. This will greatly improve the use case of the operator and make the operator the single source of truth for its certificates.
+
+To allow other components to still be able to communicate with the managed cluster via mTLS the operator will watch for a specific annotation on secrets in the namespace where an elasticsearch CR is present and will inject into this secret.
+
+### User Stories
+
+#### As an user of EO, I do not want to have to generate and provide secrets for the Elasticsearch cluster
+
+We want to further simplify using the Elasticsearch Operator and create a single source of truth for the certs the cluster uses.
+
+#### As a consumer of EO, I want a mechanism to have certificates injected into a secret for me to continue mTLS communcation with my Elasticsearch cluster
+
+We want to continue allowing mTLS as this supports the Cluster Logging and Jaeger use cases, however they shouldn't need to do more than notify they need certs injected by EO.
+
+### Implementation Details
+
+#### Assumptions
+
+* The Elasticsearch Operator will implement this in golang (rather than the current script used in CLO)
+
+* The operator will need to be able to store generated certificates for each namespace/unique cluster so that there is no security concerns of clusters cross communicating.
+
+#### Security
+
+Ensuring that signing certificates and configurations are isolated for different clusters is a must for this implementation. It will also need to be able to do this thread safe manner.
+
+### Risks and Mitigations
+
+## Design Details
+
+From a user perspective, there should be no change in any of the CRDs. The only major changes would be no longer generating secrets for EO to use, and needing to annotate a secret (user provided) in the case they want to communicate with ES in a mTLS way.
+
+### secret annotation
+
+We will add an annotation value to check for on secrets in namespaces where an es CR exists:
+
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: fluentd
+  annotations:
+    secret.openshift.io/elasticsearch-cert-component-name: system.logging.fluentd
+type: Opaque
+data: {}
+```
+
+Where the value will be optional (defaults to the name of the secret) and corresponds to the CN used for the generated certificate.
```

Comment:
We need to specify the contents of these secrets as it's part of the contract. The first decision to make is whether to keep compatibility with the current secrets. If we're keeping compatibility, the annotations need to be more expressive. Specifying just the CN is not enough, we need the names of the CA[s], the certs and the keys. Some secrets currently hold multiple copies of CA and several cert/key pairs. Please take a look at https://github.com/openshift/cluster-logging-operator/blob/c36a5589e7bc272f44153886ab1e4aba645d0638/pkg/k8shandler/certs.go#L73, this is a definition of what's currenly found in the secrets with certs as generated by the CLO.
If we're breaking compatibility with the existing format of the secrets, we can say that there's going to be a separate secret for every cert we need to generate, the name of the secret is going to be the CN unless overridden by the annotation (would need to chase down the specs to make sure any CN is representable as at least one of these), and the secret is going to contain three entries: ca, cert and key. Then there's an outlier that the kibana-proxy secret that also has an entry for its session secret, but we can find a place for it somewhere.

Reply:
i was thinking we would just have a `cert`, `key`, `ca` entry which should cover most cases. Kibana can be excluded as we wouldn't use an opt in model for it given its managed by the Elasticsearch Operator (like-wise, the ES secret wouldn't need to be opt-in)

Reply:
In lieu of these values, why not use those defined for CLF secrets.  We had a lengthy discussion in CLF before we settled on: tls.key, tls.crt, ca-bundle.crt  or something like that

Reply:
@jcantrill that's file, the document doesn't call out any specific values yet, but I can add them based on what we had in the other enhancement proposal

Reply:
After considering my ask here https://github.com/openshift/enhancements/pull/576/files#r566939322 , why does the caller who needs the secret to contact the service need to have any notion of what the CN will be?  It just needs certs that allow it to talk to the service.  As an aside, this task will be tricky because it presumably requires a cluster restart whenever a new client is added

Reply:
> As an aside, this task will be tricky because it presumably requires a cluster restart whenever a new client is added

how so? we have a signing cert that is used as the CA for the cluster, we don't need to restart at all because we'd trust that.

Reply:
> > As an aside, this task will be tricky because it presumably requires a cluster restart whenever a new client is added
> 
> how so? we have a signing cert that is used as the CA for the cluster, we don't need to restart at all because we'd trust that.

Stretching my cert limits here... is mutual auth tangential to this question since that is what we have enabled?  Wouldn't the server need to present its public key to the caller meaning it has to be in the key store?  If so could we just disable mutual auth to avoid the issue? Is it a non issue?

Reply:
this is a non-issue... we will trust the signing cert (which will be used in signing all of the client/server certs). Meaning, as long as someone trusts this original signing cert they would trust anyone that is using a cert we've signed.